### PR TITLE
Add a column to see the latest failing stage

### LINKED ISSUE #195: Add a column to see the latest failing stage
It can be useful to see at which stage a job failed at for example to see if couple of jobs failed at the same stage.

### PR DESCRIPTION
**Pull request by ciecierski**
This pr adds a column to see the latest failing stage. Closes #195 

Column has:
- a name of a failed stage 
- as an optional a link to a log file with details about the failure 

Python-jenkins in requirement file links  to the latest commit, because the latest 1.7.0 has over one year and it is missing some functions.(in my case missing get_build_stages in 1.7.0 and I didn't want to reinvent the wheel). 

Signed-off-by: ciecierski <mciecier@redhat.com>

### CODE REVIEW

File: templates/report_template.html
Comment (itzikb) [inline templates/report_template.html:85]: Shouldn't it also be N/A in case of unstable?

File: docs/README.md
Comment (nathan-weinberg) [inline docs/README.md:15]: Can this data not be gotten from the Jenkins API directly?

File: config.yaml.example
Comment (nathan-weinberg) [inline config.yaml.example:4]: If we need this field, let's put it at the bottom of the config

File: jeeves/jobs.py
Comment (nathan-weinberg) [inline jeeves/jobs.py:7]: Function looks good but why do you have a "_" at the front of the name?

File: jeeves/common.py
Comment (nathan-weinberg) [inline jeeves/common.py:54]: Since this pertains to jobs, let's have it in the jobs file and not common

File: requirements.txt
Comment (nathan-weinberg) [inline requirements.txt:3]: WRT python-jenkins - I'd really rather not have the project be using whatever is on top of master, this is messy. python-jenkins is an OpenStack project - why not reach out ot the community and ask them to draft a new release?

File: templates/report_template.html
Comment (nathan-weinberg) [inline templates/report_template.html:55]:
```suggestion
						<th style="text-align: center;">Failed Stage</th>
```

File: templates/report_template.html
Comment (nathan-weinberg) [inline templates/report_template.html:116]: If this is used multiple places, you may want to make it a macro

File: requirements.txt
Comment (itzikb) [reply requirements.txt:3]: I tried to contact of one the contributors. I'll update when I have an answer

File: jeeves/jobs.py
Comment (ciecierski) [reply jeeves/jobs.py:7]: just as a convention to declare private function

Comment (nathan-weinberg) [reply jeeves/jobs.py:7]: Let's remove it

File: templates/report_template.html
Comment (ciecierski) [reply templates/report_template.html]: Yes, you are right. I have checked it and if the state is unstable then it is problem with tests. They are already linked in the table, so I am going to set this to N/A as you suggested.

File: config.yaml.example
Comment (ciecierski) [reply config.yaml.example]: Good idea, added.

File: config.yaml.example
Comment (ciecierski) [reply config.yaml.example]: Done

File: config.yaml.example
Comment (ciecierski) [reply config.yaml.example]: Done

File: requirements.txt
Comment (nathan-weinberg) [reply requirements.txt:3]: Any word @itzikb?

Comment (ciecierski) [reply requirements.txt:3]: @itzikb  made a great job and we new pre-release in pypi https://pypi.org/project/python-jenkins/1.8.0.0a0/

Comment (nathan-weinberg) [reply requirements.txt:3]: Fantastic news!

Curious, why did they issue it as a "prerelease" instead of simply "1.8.0"?

Irregardless, let's use this tag for now

File: docs/README.md
Comment (ciecierski) [reply docs/README.md]: No, it can't be gotten from the Jenkins API directly. There is no direct API mapping between logs and stage names. However there are many common stages in infrared for all DFGs. We can try to create a basic mapping logs - stage name and put in config.example, which user can extend with DFG specific stages. Does it sound like a way to go?

**Comment by ciecierski** [comment]
> Thanks for the patch,
> I think in addition to the report template we also need the remind template

Added to remind template too

File: docs/README.md
Comment (jamepark4) [reply docs/README.md]: One thing I'm doing downstream that may be useful here is I run a tool that maps the stage to its junit xml.  I then parse the stage and failure and search through all other jobs in other DFGs to identify if the failure exists in other jobs as well.  I might be useful to have some type of extra report show cases the failures that impact you, the stage they happened, and all other DFGs that have had the same failure happen to them.  That way you have a better idea if this issue is limited to your test run or if its impacting a large number of DFGs.

Comment (odyssey4me) [reply docs/README.md]: Sounds great. I would say that should be a separate tool, though, and we can just link from the jeeves report to it.

Comment (nathan-weinberg) [reply docs/README.md]: Could also be a separate run mode, extension could be done via flag created in #192

File: templates/report_template.html
Comment (nathan-weinberg) [reply templates/report_template.html:121]: Ah cool, all good then

File: jeeves/common.py
Comment (nathan-weinberg) [reply jeeves/common.py]: Still needs to be moved

File: docs/README.md
Comment (ciecierski) [reply docs/README.md]: @nathan-weinberg I am all for separate run mode for detecting same failures in multiple DFGs jobs, however I think I can use @jamepark4  mapping function in my code instead hard-coding stage to log files. Do you have code already you can share? @jamepark4

**Review by itzikb — COMMENTED** [review]
LGTM

**Comment by nathan-weinberg** [comment]
Would like to see a section added to the README on how to use this feature as a subsection under "Usage" similar to "Filtering Builds" and "Reminder Mode"

File: docs/README.md
Comment (nathan-weinberg) [inline docs/README.md:64]:
```suggestion
Jeeves has an option to add URLs to a report which point to one of more log files for a corresponding failed build stage. Log files are mapped to the stage based on the `stage_logs` dict defined in `config.yaml`. In 
`config.yaml.example` you can find some stages from Infrared already mapped to logs files. Take it as a reference on how to map logs to your reports. A single stage can be mapped to single or multiple log files. It is not required to map all stages from a job to log files. Jeeves will skip adding URLs to stages which are not defined in the `stage_logs`.
```
Some minor changes, I also assume the last sentence is missing a "not" - let me know if that assumption is correct

File: jeeves/remind.py
Comment (nathan-weinberg) [inline jeeves/remind.py:90]:
```suggestion
						stage_urls = generate_failure_stage_log_urls(
							config, jenkins_api_info['stage_failure'],
							jenkins_api_info['job_url'], jenkins_api_info['lcb_num']
						)
```

File: templates/report_template.html
Comment (nathan-weinberg) [inline templates/report_template.html:121]: What will show here if row.stage_urls is none?

File: docs/README.md
Comment (jamepark4) [reply docs/README.md]: @ciecierski apologies I was trying to highlight the parsing and grouping a failures between all DFGs with my comment. The mapping is done in the same way you describe it, i.e. dictionary map with search patterns to relevant stages.  I can take a look to see if there are ways to dynamically map stage names to a corresponding log path, but when I wrote the code I did not see any great methods to do that.

File: templates/report_template.html
Comment (ciecierski) [reply templates/report_template.html:121]: > What will show here if row.stage_urls is none?
Nothing, no urls will be added. This column will contain only stage_name.

Comment (nathan-weinberg) [reply templates/report_template.html:121]: So, a blank square?

Comment (nathan-weinberg) [reply templates/report_template.html:121]: Discussed on IRC

**Comment by ciecierski** [comment]
failure_column
Various scenarios, how column look like visually

**Comment by nathan-weinberg** [comment]
> failure_column
> Various scenarios, how column look like visually

To match the existing styling for the rest of the report, why not have the text itself be plain or a hyperlink depending on whether or not the stage url is present?

**Comment by ciecierski** [comment]
> > failure_column
> > Various scenarios, how column look like visually
> 
> To match the existing styling for the rest of the report, why not have the text itself be plain or a hyperlink depending on whether or not the stage url is present?

Oh ,there was missing scenario, which should explain the reason. Some stages can have multiple urls. Then it can't be under stage name itself. 
columne_url_2

**Comment by nathan-weinberg** [comment]
> > > failure_column
> > > Various scenarios, how column look like visually
> > 
> > 
> > To match the existing styling for the rest of the report, why not have the text itself be plain or a hyperlink depending on whether or not the stage url is present?
> 
> Oh ,there was missing scenario, which should explain the reason. Some stages can have multiple urls. Then it can't be under stage name itself.
> columne_url_2

Ah gotcha, nvm then

File: jeeves/report.py
Comment (nathan-weinberg) [inline jeeves/report.py:158]:
```suggestion
				stage_urls = generate_failure_stage_log_urls(
					config, jenkins_api_info['stage_failure'],
					jenkins_api_info['job_url'],
					jenkins_api_info['lcb_num']
				)
```

File: jeeves/jobs.py
Comment (nathan-weinberg) [inline jeeves/jobs.py:46]: Shouldn't this only be FAILURE?

Comment (ciecierski) [reply jeeves/jobs.py]: Yes, removed it and moved also this after filter condition to get stage failures for filtered jobs.

File: docs/README.md
Comment (nathan-weinberg) [inline docs/README.md:25]:
```suggestion
- **stage_logs**: Optional dict field that instructs Jeeves how to build URLs to a log files for a corresponding failed build stage
```

File: docs/README.md
Comment (nathan-weinberg) [inline docs/README.md:64]:
```suggestion
`config.yaml.example` you can find some stages already mapped to logs files. Use it as a reference on how to map logs to your reports. A single stage can be mapped to single or multiple log files. It is not required to map all stages from a job to log files. Jeeves will skip adding URLs to stages which are not defined in the `stage_logs`.
```

File: jeeves/jobs.py
Comment (nathan-weinberg) [inline jeeves/jobs.py:9]:
```suggestion
		returns string with the name of failed stage or 'N/A' if there is no failed stage
```

**Review by nathan-weinberg — CHANGES_REQUESTED** [review]
Some stylistic requests, also per my email reminders are not working

@agarcia64 @amodi5 @jamepark4 @pbabbar3 Please test reports/reminders for your respective DFGs and report back when you can, thanks

**Review by nathan-weinberg — APPROVED** [review]
Worked with Enterprise report and reminders
Would like to see more DFGs testing this, but given how long this PR has been open I will merge this on Monday regardless

**Comment by jamepark4** [comment]
> Worked with Enterprise report and reminders
> Would like to see more DFGs testing this, but given how long this PR has been open I will merge this on Monday regardless

Just tested Compute report and no issues for me.

**Review by agarcia64 — CHANGES_REQUESTED** [review]
Does not work when job_search_fields in the config.yaml includes 'DFG-ceph-rhos'. Specifically the red, failing jobs will appear as grey and do not include the failing stage.

**Comment by nathan-weinberg** [comment]
@ciecierski @agarcia64 Please work together to fix the issue with the Ceph jobs before we merge

**Review by agarcia64 — APPROVED** [review]
The problem I was facing went away after I installed the new requirements.txt. Sorry for holding up the merging, everything is working as expected on my end and the code LGTM.

**Pull request merged**
